### PR TITLE
Updates/pkg hardware

### PR DESCRIPTION
**Pull request by elvece**
Actually quite a massive PR as this change affected a lot. The main logic is in fetching the new query params, and `filterDevices`. There is some refactor with changes to the DB model.

The first commit is for the changes to the publish script for adding/removing arches for a package. The second commit is the API changes. Other commits are fixes.

This is now well tested with no known issues and full backwards compatibility.

### CODE REVIEW

File: src/Cli/Cli.hs [hunk 1]
```diff
@@ -281,15 +284,25 @@ parsePublish =
...
                 ( optional $
                     strOption
                         (short 'p' <> long "package" <> metavar "S9PK" <> help "File path of the package to publish")
                 )
                 (switch (short 'i' <> long "index" <> help "Index the package after uploading"))
+                ( optional $
+                    some parseArch 
+                )
 
+parseArch :: Parser OsArch
```
Comment (Blu-J) [inline src/Cli/Cli.hs:299]: :+1:

File: src/Database/Queries.hs [hunk 2]
```diff
@@ -326,17 +295,34 @@ upsertPackageVersion PackageManifest{..} = do
...
+getVersionPlatform ::
+    (Monad m, MonadIO m) =>
+    PkgRecordId ->
+    [Maybe OsArch] ->
+    ReaderT SqlBackend m [VersionPlatform]
+getVersionPlatform pkgId arches = do
+    vps <- select $ do
+        v <- from $ table @VersionPlatform
+        where_ $ v ^. VersionPlatformPkgId ==. val pkgId
+        where_ (v ^. VersionPlatformArch `in_` valList arches)
```
Comment (Blu-J) [inline src/Database/Queries.hs:326]: Why 
`fn $ args args 2`
vs 
`fn (args args 2)`

File: src/Database/Queries.hs [hunk 3]
```diff
@@ -326,17 +295,34 @@ upsertPackageVersion PackageManifest{..} = do
...
+        pure v
```
Comment (Blu-J) [inline src/Database/Queries.hs:327]: :question: So, is the where mutating v?

**Review by Blu-J — APPROVED** [review]

File: src/Handler/Package/V1/Index.hs [hunk 4]
```diff
@@ -172,9 +185,29 @@ getLimitQuery :: Handler (Maybe Int)
...
+    osVersion <- getOsVersionCompatQuery >>= \case
+        Just a -> pure $ Just a
+        Nothing -> getOsVersionCompatQueryLegacy
+    pure osVersion
+
+getHardwareDevicesQuery :: Handler (HM.HashMap Text Text)
+getHardwareDevicesQuery = do
+    allParams <- reqGetParams <$> getRequest
+    -- [("hardware.device.processor","intel"),("hardware.device.display","led")]
+    let hardwareDeviceParams = filter (\(key, _) -> "hardware.device" `isInfixOf` key) allParams
```
Comment (dr-bonez) [inline src/Handler/Package/V1/Index.hs:205]:
```suggestion
    let hardwareDeviceParams = filter (\(key, _) -> "hardware.device." `isPrefixOf` key) allParams
```

File: src/Lib/Types/Manifest.hs [hunk 5]
```diff
@@ -79,6 +109,13 @@ testManifest =
     "short": "Create Tor websites, hosted on your Embassy.",
     "long": "Embassy Pages is a simple web server that uses directories inside File Browser to serve Tor websites."
   },
+  "hardware-requirements" {
+    "device": {
+      "processor": "intel",
+      "display": "r'^{.*}$'"
+    },
+    "ram": "8"
```
Comment (dr-bonez) [inline src/Lib/Types/Manifest.hs:117]: this package requires at least 8 whole bytes of ram! Wow!

Comment (elvece) [reply src/Lib/Types/Manifest.hs:117]: 😅 also i have fixed that both keys are regex

**Pull request merged**